File: data/src/main/java/com/github/anrimian/musicplayer/data/controllers/music/players/ExoMediaPlayer.java
package com.github.anrimian.musicplayer.data.controllers.music.players;

import android.content.Context;
import android.net.Uri;

import com.github.anrimian.musicplayer.data.controllers.music.equalizer.EqualizerController;
import com.github.anrimian.musicplayer.data.controllers.music.error.PlayerErrorParser;
import com.github.anrimian.musicplayer.data.controllers.music.players.exoplayer.StereoVolumeProcessor;
import com.github.anrimian.musicplayer.data.models.composition.source.UriCompositionSource;
import com.github.anrimian.musicplayer.data.storage.source.CompositionSourceProvider;
import com.github.anrimian.musicplayer.data.utils.exo_player.PlayerEventListener;
import com.github.anrimian.musicplayer.domain.models.composition.source.CompositionSource;
import com.github.anrimian.musicplayer.domain.models.composition.source.LibraryCompositionSource;
import com.github.anrimian.musicplayer.domain.models.player.SoundBalance;
import com.github.anrimian.musicplayer.domain.models.player.error.ErrorType;
import com.github.anrimian.musicplayer.domain.models.player.events.ErrorEvent;
import com.github.anrimian.musicplayer.domain.models.player.events.FinishedEvent;
import com.github.anrimian.musicplayer.domain.models.player.events.PlayerEvent;
import com.github.anrimian.musicplayer.domain.models.player.events.PreparedEvent;
import com.github.anrimian.musicplayer.domain.utils.functions.Callback;
import com.google.android.exoplayer2.ExoPlayer;
import com.google.android.exoplayer2.DefaultRenderersFactory;
import com.google.android.exoplayer2.ExoPlaybackException;
import com.google.android.exoplayer2.MediaItem;
import com.google.android.exoplayer2.PlaybackException;
import com.google.android.exoplayer2.PlaybackParameters;
import com.google.android.exoplayer2.RenderersFactory;
import com.google.android.exoplayer2.SimpleExoPlayer;
import com.google.android.exoplayer2.audio.AudioCapabilities;
import com.google.android.exoplayer2.audio.AudioProcessor;
import com.google.android.exoplayer2.audio.AudioSink;
import com.google.android.exoplayer2.audio.DefaultAudioSink;
import com.google.android.exoplayer2.source.MediaSource;
import com.google.android.exoplayer2.source.ProgressiveMediaSource;
import com.google.android.exoplayer2.upstream.DataSource;
import com.google.android.exoplayer2.upstream.DefaultDataSource;
import com.google.android.exoplayer2.upstream.Loader;

import java.util.concurrent.TimeUnit;

import javax.annotation.Nullable;

import io.reactivex.rxjava3.core.Completable;
import io.reactivex.rxjava3.core.Observable;
import io.reactivex.rxjava3.core.Scheduler;
import io.reactivex.rxjava3.core.Single;
import io.reactivex.rxjava3.disposables.Disposable;
import io.reactivex.rxjava3.subjects.BehaviorSubject;
import io.reactivex.rxjava3.subjects.PublishSubject;

public class ExoMediaPlayer implements AppMediaPlayer {

    private final BehaviorSubject<Long> trackPositionSubject = BehaviorSubject.create();
    private final PublishSubject<PlayerEvent> playerEventSubject = PublishSubject.create();

    private final Context context;
    private final CompositionSourceProvider sourceRepository;
    private final Scheduler uiScheduler;
    private final Scheduler ioScheduler;
    private final PlayerErrorParser playerErrorParser;
    private final EqualizerController equalizerController;

    private final StereoVolumeProcessor stereoVolumeProcessor = new StereoVolumeProcessor();

    private volatile ExoPlayer player;

    @Nullable
    private Disposable trackPositionDisposable;

    private CompositionSource currentComposition;

    private boolean isPreparing = false;
    private boolean playAfterPrepare = false;

    public ExoMediaPlayer(Context context,
                          CompositionSourceProvider sourceRepository,
                          Scheduler uiScheduler,
                          Scheduler ioScheduler,
                          PlayerErrorParser playerErrorParser,
                          EqualizerController equalizerController) {
        this.context = context;
        this.playerErrorParser = playerErrorParser;
        this.sourceRepository = sourceRepository;
        this.uiScheduler = uiScheduler;
        this.ioScheduler = ioScheduler;
        this.equalizerController = equalizerController;

        stereoVolumeProcessor.setChannelMap(new int[] { 0, 1 } );
    }

    @Override
    public Observable<PlayerEvent> getEventsObservable() {
        return playerEventSubject;
    }

    @Override
    public void prepareToPlay(CompositionSource composition,
                              long startPosition,
                              @Nullable ErrorType previousErrorType) {
        isPreparing = true;
        this.currentComposition = composition;
        trackPositionSubject.onNext(startPosition);
        //cancel previous preparation?
        Single.fromCallable(() -> composition)
                .flatMapCompletable(this::prepareMediaSource)
                .doOnEvent(t -> onCompositionPrepared(t, startPosition))
                .onErrorComplete()
                .subscribeOn(uiScheduler)
                .subscribe();
    }

    @Override
    public void stop() {
        Completable.fromRunnable(() -> {
            seekTo(0);
            pausePlayer();
            stopTracingTrackPosition();
        }).subscribeOn(uiScheduler).subscribe();
    }

    @Override
    public void resume() {
        if (isPreparing) {
            playAfterPrepare = true;
            return;
        }
        startPlayWhenReady();
    }

    @Override
    public void pause() {
        Completable.fromRunnable(() -> {
            pausePlayer();
            stopTracingTrackPosition();
        }).subscribeOn(uiScheduler).subscribe();
    }

    @Override
    public void seekTo(long position) {
        Completable.fromRunnable(() -> {
            try {
                getPlayer().seekTo(position);
            } catch (IndexOutOfBoundsException ignored) {//crash inside exoplayer
                return;
            }
            trackPositionSubject.onNext(position);
        }).subscribeOn(uiScheduler).subscribe();
    }

    @Override
    public void setVolume(float volume) {
        Completable.fromRunnable(() -> getPlayer().setVolume(volume))
                .subscribeOn(uiScheduler)
                .subscribe();
    }

    @Override
    public Observable<Long> getTrackPositionObservable() {
        return trackPositionSubject;
    }

    @Override
    public Single<Long> getTrackPosition() {
        return Single.fromCallable(() -> getPlayer().getCurrentPosition())
                .subscribeOn(uiScheduler);
    }

    @Override
    public Single<Long> seekBy(long millis) {
        return getTrackPosition()
                .map(currentPosition -> {
                    long targetPosition = currentPosition + millis;
                    if (targetPosition < 0) {
                        targetPosition = 0;
                    }
                    if (targetPosition > getPlayer().getDuration()) {
                        return currentPosition;
                    }
                    seekTo(targetPosition);
                    return targetPosition;
                });
    }

    @Override
    public void setPlaySpeed(float speed) {
        usePlayer(player -> {
            PlaybackParameters param = new PlaybackParameters(speed);
            player.setPlaybackParameters(param);
        });
    }

    @Override
    public void release() {
        usePlayer(player -> {
//            equalizerController.detachEqualizer();
            pausePlayer();
            stopTracingTrackPosition();
            player.release();
        });
    }

    @Override
    public Observable<Boolean> getSpeedChangeAvailableObservable() {
        return Observable.fromCallable(() -> true);
    }

    @Override
    public void setSoundBalance(SoundBalance soundBalance) {
        stereoVolumeProcessor.setVolume(soundBalance.getLeft(), soundBalance.getRight());
    }

    private void startPlayWhenReady() {
        Completable.fromRunnable(() -> {
            getPlayer().setPlayWhenReady(true);
            equalizerController.attachEqualizer(getPlayer().getAudioSessionId());
            startTracingTrackPosition();
        }).subscribeOn(uiScheduler).subscribe();
    }

    private void onCompositionPrepared(Throwable throwable, long startPosition) {
        isPreparing = false;
        if (throwable == null) {
            seekTo(startPosition);
            playerEventSubject.onNext(new PreparedEvent(currentComposition));
            if (playAfterPrepare) {
                playAfterPrepare = false;
                startPlayWhenReady();
            }
        } else {
            seekTo(0);
            pausePlayer();
            sendErrorEvent(throwable);
        }
    }

    private void pausePlayer() {
        getPlayer().setPlayWhenReady(false);
        equalizerController.detachEqualizer();
    }

    private void sendErrorEvent(Throwable throwable) {
        if (currentComposition != null) {
            //workaround for prepareError in newest exo player versions
            if (isStrangeLoaderException(throwable)) {
                prepareToPlay(currentComposition, getPlayer().getCurrentPosition(), null);
                return;
            }

            playerEventSubject.onNext(new ErrorEvent(
                    playerErrorParser.getErrorType(throwable),
                    currentComposition)
            );
        }
    }

    private void startTracingTrackPosition() {
        stopTracingTrackPosition();
        trackPositionDisposable = Observable.interval(0, 1, TimeUnit.SECONDS)
                .observeOn(uiScheduler)
                .map(o -> getPlayer().getCurrentPosition())
                .subscribe(trackPositionSubject::onNext);
    }

    private void stopTracingTrackPosition() {
        if (trackPositionDisposable != null) {
            trackPositionDisposable.dispose();
            trackPositionDisposable = null;
        }
    }

    private Completable prepareMediaSource(CompositionSource composition) {
        return getCompositionUri(composition)
                .flatMap(this::createMediaSource)
                .subscribeOn(ioScheduler)
                .timeout(6, TimeUnit.SECONDS)//read from uri can be freeze for some reason, check
                .observeOn(uiScheduler)
                .doOnSuccess(mediaSource -> {
                    getPlayer().setMediaSource(mediaSource);
                    getPlayer().prepare();
                })
                .ignoreElement();
    }

    private Single<Uri> getCompositionUri(CompositionSource composition) {
        if (composition instanceof LibraryCompositionSource) {
            long id = ((LibraryCompositionSource) composition).getComposition().getId();
            return sourceRepository.getCompositionUri(id);
        }
        if (composition instanceof UriCompositionSource) {
            return Single.fromCallable(((UriCompositionSource) composition)::getUri);
        }
        throw new IllegalArgumentException("unknown composition source");
    }

    private boolean isStrangeLoaderException(Throwable throwable) {
        if (throwable instanceof PlaybackException) {
            Throwable cause = throwable.getCause();
            return cause instanceof Loader.UnexpectedLoaderException;
        }
        return false;
    }

    private Single<MediaSource> createMediaSource(Uri uri) {
        return Single.fromCallable(() -> {
            DataSource.Factory dataSourceFactory = new DefaultDataSource.Factory(context);
            MediaItem mediaItem = MediaItem.fromUri(uri);
            return new ProgressiveMediaSource.Factory(dataSourceFactory).createMediaSource(mediaItem);
        });
    }

    private void usePlayer(Callback<ExoPlayer> function) {
        Completable.fromAction(() -> function.call(getPlayer()))
                .subscribeOn(uiScheduler)
                .subscribe();
    }

    private ExoPlayer getPlayer() {
        if (player == null) {
            synchronized (this) {
                if (player == null) {
                    RenderersFactory factory = createSimpleRenderersFactory(context, stereoVolumeProcessor);

                    player = new ExoPlayer.Builder(context, factory)
                            .build();

                    PlayerEventListener playerEventListener = new PlayerEventListener(
                            () -> playerEventSubject.onNext(new FinishedEvent(currentComposition)),
                            this::sendErrorEvent
                    );
                    player.addListener(playerEventListener);
//                    equalizerController.attachEqualizer(player.getAudioSessionId());
//                    player.addAnalyticsListener(new AnalyticsListener() {
//
//                        @Override
//                        public void onAudioSessionIdChanged(@NonNull EventTime eventTime, int audioSessionId) {
//                            equalizerController.attachEqualizer(audioSessionId);
//                        }

//                    });

                }
            }
        }
        return player;
    }

    private RenderersFactory createSimpleRenderersFactory(Context context,
                                                          AudioProcessor... audioProcessors) {
        return new DefaultRenderersFactory(context) {

            @Override
            protected AudioSink buildAudioSink(Context context1,
                                               boolean enableFloatOutput,
                                               boolean enableAudioTrackPlaybackParams,
                                               boolean enableOffload) {
                return new DefaultAudioSink(
                        AudioCapabilities.getCapabilities(context1),
                        new DefaultAudioSink.DefaultAudioProcessorChain(audioProcessors),
                        enableFloatOutput,
                        enableAudioTrackPlaybackParams,
                        enableOffload
                                ? DefaultAudioSink.OFFLOAD_MODE_ENABLED_GAPLESS_REQUIRED
                                : DefaultAudioSink.OFFLOAD_MODE_DISABLED);
            }
        };
    }

}

File: data/src/main/java/com/github/anrimian/musicplayer/data/controllers/music/players/ExoMediaPlayer.kt
package com.github.anrimian.musicplayer.data.controllers.music.players

import android.annotation.SuppressLint
import android.content.Context
import android.net.Uri
import androidx.media3.common.AudioAttributes
import androidx.media3.common.C
import androidx.media3.common.MediaItem
import androidx.media3.common.PlaybackException
import androidx.media3.common.PlaybackParameters
import androidx.media3.common.audio.AudioProcessor
import androidx.media3.common.audio.SonicAudioProcessor
import androidx.media3.datasource.DefaultDataSource
import androidx.media3.exoplayer.DefaultRenderersFactory
import androidx.media3.exoplayer.ExoPlayer
import androidx.media3.exoplayer.RenderersFactory
import androidx.media3.exoplayer.audio.AudioSink
import androidx.media3.exoplayer.audio.DefaultAudioSink
import androidx.media3.exoplayer.audio.SilenceSkippingAudioProcessor
import androidx.media3.exoplayer.source.MediaSource
import androidx.media3.exoplayer.source.ProgressiveMediaSource
import androidx.media3.exoplayer.source.UnrecognizedInputFormatException
import androidx.media3.exoplayer.upstream.Loader
import androidx.media3.extractor.DefaultExtractorsFactory
import androidx.media3.extractor.ts.AdtsExtractor
import com.github.anrimian.musicplayer.data.controllers.music.equalizer.EqualizerController
import com.github.anrimian.musicplayer.data.controllers.music.players.exceptions.PlayerOutOfMemoryException
import com.github.anrimian.musicplayer.data.controllers.music.players.exoplayer.StereoVolumeProcessor
import com.github.anrimian.musicplayer.data.controllers.music.players.utils.ExoPlayerMediaItemBuilder
import com.github.anrimian.musicplayer.data.utils.exo_player.PlayerEventListener
import com.github.anrimian.musicplayer.domain.models.composition.content.CompositionContentSource
import com.github.anrimian.musicplayer.domain.models.composition.content.NoReadPermissionException
import com.github.anrimian.musicplayer.domain.models.composition.content.UnsupportedSourceException
import com.github.anrimian.musicplayer.domain.models.player.SoundBalance
import com.github.anrimian.musicplayer.domain.models.player.events.MediaPlayerEvent
import io.reactivex.rxjava3.core.Completable
import io.reactivex.rxjava3.core.Observable
import io.reactivex.rxjava3.core.Scheduler
import io.reactivex.rxjava3.core.Single
import io.reactivex.rxjava3.subjects.PublishSubject
import java.util.concurrent.TimeUnit

@SuppressLint("UnsafeOptInUsageError")
class ExoMediaPlayer(
    private val context: Context,
    private val uiScheduler: Scheduler,
    private val equalizerController: EqualizerController,
    private val exoPlayerMediaItemBuilder: ExoPlayerMediaItemBuilder,
    private val hasAlternativePlayer: Boolean,
) : AppMediaPlayer {

    private val playerEventsSubject = PublishSubject.create<MediaPlayerEvent>()
    private val stereoVolumeProcessor = StereoVolumeProcessor().apply {
        setChannelMap(intArrayOf(0, 1))
    }
    private val silenceSkippingAudioProcessor = SilenceSkippingAudioProcessor(
        /* minimumSilenceDurationUs = */ 2000_000L,
        /* silenceRetentionRatio = */ 0f,
        /* maxSilenceToKeepDurationUs = */ 2000_000L,
        /* minVolumeToKeepPercentageWhenMuting = */ 0,
        /* silenceThresholdLevel = */ 128.toShort()
    )

    private val player by lazy {
        val factory = createSimpleRenderersFactory(
            context,
            stereoVolumeProcessor,
            silenceSkippingAudioProcessor
        )
        val player = ExoPlayer.Builder(context, factory).build()
        val audioAttributes = AudioAttributes.Builder()
            .setUsage(C.USAGE_MEDIA)
            .setContentType(C.AUDIO_CONTENT_TYPE_MUSIC)
            .build()
        player.setAudioAttributes(audioAttributes, false)
        val playerEventListener = PlayerEventListener(
            { playerEventsSubject.onNext(MediaPlayerEvent.Finished) },
            { t -> playerEventsSubject.onNext(MediaPlayerEvent.Error(mapPlayerException(t)))})
        player.addListener(playerEventListener)
        return@lazy player
    }

    override fun getPlayerEventsObservable(): Observable<MediaPlayerEvent> {
        return playerEventsSubject
    }

    override fun prepareToPlay(
        source: CompositionContentSource,
        previousException: Exception?,
    ): Completable {
        return Single.just(exoPlayerMediaItemBuilder.createUri(source))
            .doOnSuccess { uri ->
                if (hasAlternativePlayer && isRawAacFile(uri)) {
                    throw UnsupportedSourceException()
                }
            }
            .flatMap(this::createMediaSource)
            .observeOn(uiScheduler)
            .doOnSuccess { mediaSource ->
                player.setMediaSource(mediaSource)
                player.prepare()
            }
            .ignoreElement()
            .onErrorResumeNext { t -> Completable.error(mapPlayerException(t)) }
    }

    override fun stop() {
        Completable.fromRunnable {
            seekTo(0)
            pausePlayer()
        }.subscribeOn(uiScheduler).subscribe()
    }

    override fun resume() {
        startPlayWhenReady()
    }

    override fun pause() {
        Completable.fromRunnable {
            pausePlayer()
        }.subscribeOn(uiScheduler).subscribe()
    }

    override fun seekTo(position: Long) {
        Completable.fromRunnable {
            try {
                player.seekTo(position)
            } catch (_: IndexOutOfBoundsException) { //crash inside exoplayer
                return@fromRunnable
            }
        }.subscribeOn(uiScheduler).subscribe()
    }

    override fun setVolume(volume: Float) {
        Completable.fromRunnable { player.volume = volume }
            .subscribeOn(uiScheduler)
            .subscribe()
    }

    override fun getTrackPositionObservable(): Observable<Long> {
        return Observable.interval(0, 50, TimeUnit.MILLISECONDS)
            .observeOn(uiScheduler)
            .map { player.currentPosition }
    }

    override fun getTrackPosition(): Single<Long> {
        return Single.fromCallable { player.currentPosition }
            .subscribeOn(uiScheduler)
    }

    override fun getDuration(): Single<Long> {
        return Single.fromCallable { player.duration }
            .subscribeOn(uiScheduler)
    }

    override fun setPlaybackSpeed(speed: Float) {
        usePlayer { player -> player.playbackParameters = PlaybackParameters(speed) }
    }

    override fun setSkipSilenceEnabled(enabled: Boolean) {
        usePlayer { player -> player.skipSilenceEnabled = enabled }
    }

    override fun release() {
        usePlayer { player ->
//            equalizerController.detachEqualizer();
            pausePlayer()
            player.release()
        }
    }

    override fun getSpeedChangeAvailableObservable(): Observable<Boolean> {
        return Observable.fromCallable { true }
    }

    override fun setSoundBalance(soundBalance: SoundBalance) {
        stereoVolumeProcessor.setVolume(soundBalance.left, soundBalance.right)
    }

    private fun startPlayWhenReady() {
        Completable.fromRunnable {
            player.playWhenReady = true
            equalizerController.attachEqualizer(player.audioSessionId)
        }.subscribeOn(uiScheduler).subscribe()
    }

    private fun pausePlayer() {
        player.playWhenReady = false
        equalizerController.detachEqualizer()
    }

    private fun mapPlayerException(throwable: Throwable): Throwable {
        //logic is duplicated in PlayerErrorParserImpl
        //likely all exo player error parsing logic should be here
        if (throwable is PlaybackException) {
            when (val cause = throwable.cause) {
                is UnrecognizedInputFormatException -> {
                    return UnsupportedSourceException()
                }
                is Loader.UnexpectedLoaderException -> {
                    when (val causeOfCause = cause.cause) {
                        is OutOfMemoryError -> return PlayerOutOfMemoryException()
                        is SecurityException -> return NoReadPermissionException(causeOfCause)
                    }
                }
            }
        }
        return throwable
    }

    private fun createMediaSource(uri: Uri): Single<MediaSource> {
        return Single.fromCallable<MediaSource> {
            val dataSourceFactory = DefaultDataSource.Factory(context)
            val extractorsFactory = DefaultExtractorsFactory()
                .setAdtsExtractorFlags(AdtsExtractor.FLAG_ENABLE_CONSTANT_BITRATE_SEEKING)
            val mediaItem = MediaItem.fromUri(uri)
            ProgressiveMediaSource.Factory(dataSourceFactory, extractorsFactory)
                .createMediaSource(mediaItem)
        }.timeout(6, TimeUnit.SECONDS) //read from uri can be freeze for some reason, check
    }

    private fun isRawAacFile(uri: Uri): Boolean {
        return uri.lastPathSegment?.endsWith(".aac", ignoreCase = true) == true
    }

    private fun usePlayer(function: (ExoPlayer) -> Unit) {
        Completable.fromAction { function(player) }
            .subscribeOn(uiScheduler)
            .subscribe()
    }

    private fun createSimpleRenderersFactory(
        context: Context,
        stereoProcessor: AudioProcessor,
        silenceProcessor: SilenceSkippingAudioProcessor
    ): RenderersFactory {
        return object : DefaultRenderersFactory(context) {
            override fun buildAudioSink(
                context: Context,
                enableFloatOutput: Boolean,
                enableAudioTrackPlaybackParams: Boolean,
            ): AudioSink {

                val audioProcessorChain = DefaultAudioSink.DefaultAudioProcessorChain(
                    arrayOf(stereoProcessor),
                    silenceProcessor,
                    SonicAudioProcessor()
                )

                return DefaultAudioSink.Builder(context)
                    .setAudioProcessorChain(audioProcessorChain)
                    .setEnableFloatOutput(enableFloatOutput)
                    .setEnableAudioTrackPlaybackParams(enableAudioTrackPlaybackParams)
                    .build()
            }
        }
    }

}